src/main/java/org/wildfly/security/authz/MappedRoleMapper.java

```diff
@@ -0,0 +1,143 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.wildfly.security.authz;
+
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Set;
+
+import static org.wildfly.security._private.ElytronMessages.log;
+
+/**
+ * A simple mapping role mapper.
+ *
+ * Maps each role to a set of new roles using a String to Set<String> map.
+ *
+ * @author <a href="mailto:mmazanek@redhat.com">Martin Mazanek</a>
+ */
+public class MappedRoleMapper implements RoleMapper {
+
+    private final Map<String, Set<String>> reverseRoleMap;
+    private volatile boolean initialized = false;
+
+    /**
+     * Construct a new instance.
+     * Called from WildFly core when using this as a custom component. You should not use this constructor and use {@link MappedRoleMapper.Builder} instead.
+     *
+     * @see MappedRoleMapper.Builder
+     */
+    public MappedRoleMapper() {
+        this.reverseRoleMap = new LinkedHashMap<>();
+    }
+
+
+    private MappedRoleMapper(Map<String, Set<String>> roleMap) {
+        this.reverseRoleMap = new LinkedHashMap<>();
+
+        Set<Map.Entry<String, Set<String>>> entrySet = roleMap.entrySet();
+
+        for (Map.Entry<String, Set<String>> entry : entrySet) {
+
+            for (String mappedRole : entry.getValue()) {
+                Set<String> rolesToMappedRole = reverseRoleMap.get(mappedRole);
+
+                if (rolesToMappedRole == null) {
+                    rolesToMappedRole = new LinkedHashSet<>();
+                    reverseRoleMap.put(mappedRole, rolesToMappedRole);
+                }
+
+                rolesToMappedRole.add(entry.getKey());
+            }
+        }
+
+        initialized = true;
+    }
+
+    /**
+     * Custom component method.
+     * Called from WildFly core. Used to include mapped role mapping functionality in older WildFly versions.
+     *
+     * @param configuration map of mapping rules where key is delegate role and value is whitespace separated list of new roles
+     */
+    public void initialize(final Map<String, String> configuration) {
+        if (initialized) {
+            throw log.roleMappedAlreadyInitialized();
+        }
+        reverseRoleMap.clear();
+        configuration.forEach( (key, value) -> {
+            String[] newRoles = value.split("\\s+");
+            for (String newRole : newRoles) {
+                Set<String> rolesToMappedRole = reverseRoleMap.get(newRole);
+                if (rolesToMappedRole == null) {
+                    rolesToMappedRole = new LinkedHashSet<>();
+                    reverseRoleMap.put(newRole, rolesToMappedRole);
+                }
+                rolesToMappedRole.add(key);
+            }
+        });
+        initialized = true;
```

Comment:
As the object should be initialized before it start to be shared between threads, **it should be ok** initialized is not set+get atomically.